pkg/controller/iotconfig/auth_service.go

```diff
@@ -0,0 +1,226 @@
+		return r.processService(ctx, nameAuthService, config, r.reconcileAuthServiceService)
+	})
+	rc.ProcessSimple(func() error {
+		return r.processConfigMap(ctx, nameAuthService+"-config", config, r.reconcileAuthServiceConfigMap)
+	})
+
+	return rc.Result()
+}
+
+func (r *ReconcileIoTConfig) reconcileAuthServiceDeployment(config *iotv1alpha1.IoTConfig, deployment *appsv1.Deployment) error {
+
+	install.ApplyDeploymentDefaults(deployment, "iot", appAuthService, deployment.Name)
+
+	deployment.Spec.Replicas = nil
+
+	err := install.ApplyContainerWithError(deployment, "auth-service", func(container *corev1.Container) error {
+		if err := install.SetContainerImage(container, "iot-auth-service", config); err != nil {
+			return err
+		}
+
+		container.Resources = corev1.ResourceRequirements{
+			Limits: corev1.ResourceList{
+				corev1.ResourceMemory: *resource.NewQuantity(512*1024*1024 /* 512Mi */, resource.BinarySI),
+			},
+		}
+		container.Ports = []corev1.ContainerPort{
+			{Name: "jolokia", ContainerPort: 8778, Protocol: corev1.ProtocolTCP},
+			{Name: "amqps", ContainerPort: 5671, Protocol: corev1.ProtocolTCP},
+		}
+
+		SetHonoProbes(container)
+
+		// environment
+
+		container.Env = []corev1.EnvVar{
+			{Name: "SPRING_CONFIG_LOCATION", Value: "file:///etc/config/"},
+			{Name: "SPRING_PROFILES_ACTIVE", Value: "authentication-impl"},
+			{Name: "LOGGING_CONFIG", Value: "file:///etc/config/logback-spring.xml"},
+			{Name: "KUBERNETES_NAMESPACE", ValueFrom: &corev1.EnvVarSource{FieldRef: &corev1.ObjectFieldSelector{FieldPath: "metadata.namespace"}}},
+		}
+
+		// volume mounts
+
+		if len(container.VolumeMounts) != 2 {
+			container.VolumeMounts = make([]corev1.VolumeMount, 2)
+		}
+
+		container.VolumeMounts[0].Name = "conf"
+		container.VolumeMounts[0].MountPath = "/etc/config"
+		container.VolumeMounts[0].ReadOnly = false
+
+		container.VolumeMounts[1].Name = "tls"
+		container.VolumeMounts[1].MountPath = "/etc/tls"
+		container.VolumeMounts[1].ReadOnly = true
+
+		// return
+
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// volumes
+
+	install.ApplyConfigMapVolume(deployment, "conf", nameAuthService+"-config")
+	install.ApplySecretVolume(deployment, "tls", "iot-auth-service-tls")
+
+	// return
+
+	return nil
+}
+
+func (r *ReconcileIoTConfig) reconcileAuthServiceService(config *iotv1alpha1.IoTConfig, service *corev1.Service) error {
+
+	install.ApplyServiceDefaults(service, "iot", appAuthService, service.Name)
+
+	if len(service.Spec.Ports) != 1 {
+		service.Spec.Ports = make([]corev1.ServicePort, 1)
+	}
+
+	service.Spec.Ports[0].Name = "amqps"
+	service.Spec.Ports[0].Port = 5671
+	service.Spec.Ports[0].TargetPort = intstr.FromInt(5671)
+	service.Spec.Ports[0].Protocol = corev1.ProtocolTCP
+
+	if service.Annotations == nil {
+		service.Annotations = make(map[string]string)
+	}
+
+	// FIXME: remove OpenShift specific feature
+	service.Annotations["service.alpha.openshift.io/serving-cert-secret-name"] = nameAuthService + "-tls"
```

Comment:
FWIW, one approach I'm taking in the authservice controller is to allow a secret with the certificate to be specified in the CRD. On Kubernetes, this is the only way to get the cert, whereas on OpenShift, we can 'default' to this annotation.

Reply:
Yes, my idea was to fix this by generating a certificate in the controller, but simply providing this might be a good alternative for that.

Reply:
> On Kubernetes, this is the only way to get the cert, …

I think one could also use the operator on plain k8s … https://github.com/openshift/service-ca-operator

Would it be sufficient to provide the name of the secret in the CRD? Or do you want to have the full content?

Reply:
I've used the `SecretReference` type for this information. I.e.

```
type AuthenticationServiceSpecNone struct {
  CertificateSecret *corev1.SecretReference       `json:"certificateSecret"`
  ...
}
```

Reply:
This allows using a separate namespace, do we want to support that?

Reply:
As discussed, we will do this is a follow up PR.